### PR TITLE
OSDOCS-8895: introduce network config options and limits

### PR DESCRIPTION
**Pull request by ShaunaDiaz**
Version(s):
4.14+

Issue:
https://issues.redhat.com/browse/OSDOCS-8895

Link to docs preview:
[About the CNI](https://file.rdu.redhat.com/shdiaz/OSDOCS-8895/microshift_networking/microshift-cni.html)
[Feature matrix](https://file.rdu.redhat.com/shdiaz/OSDOCS-8895/microshift_networking/microshift-cni.html#microshift-nw-customization-matrix_microshift-about-ovn-k-plugin)

QE review:
- [x] QE has approved this change.

### CODE REVIEW

**Comment by ShaunaDiaz** [comment]
@pacevedom PTAL at this first issue in our networking series. I have two questions, too, if you look at the diffs.

File: modules/microshift-cni-customization-matrix.adoc [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+|Firewall|No ^[1]^|Supported
+
+|Hardware offloading|Not available|N/A
+
+|Hybrid networking|Not available|N/A
+
+|IPsec encryption for intra-cluster communication|Not available|N/A
+
+|IPv6||Not available|N/A
```
Comment (kelbrown20) [inline modules/microshift-cni-customization-matrix.adoc:37]: Just to fix the format
```suggestion
|IPv6|Not available|N/A
```

File: modules/microshift-cni-customization-matrix.adoc [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+|Kubernetes network policy logs|Not available|N/A
+|===
+
+1. Setting up a firewall is supported by the OVN-K CNI.
```
Comment (kelbrown20) [inline modules/microshift-cni-customization-matrix.adoc:42]: I think these and the numbers in the table should be switched just to be sequential cuz in the table rn, 2 comes first

File: microshift_networking/microshift-cni.adoc [hunk 3]
```diff
@@ -6,7 +6,22 @@ include::_attributes/attributes-microshift.adoc[]
...
 
-OVN-Kubernetes is the default networking solution for {microshift-short} deployments. OVN-Kubernetes is a virtualized network for pods and services that is based on Open Virtual Network (OVN). The OVN-Kubernetes Container Network Interface (CNI) plugin is the network plugin for the cluster. A cluster that uses the OVN-Kubernetes network plugin also runs Open vSwitch (OVS) on the node. OVN configures OVS on the node to implement the declared network configuration.
+The OVN-Kubernetes Container Network Interface (CNI) plugin is the default networking solution for {microshift-short} clusters. OVN-Kubernetes is a virtualized network for pods and services that is based on Open Virtual Network (OVN).
+
+* Default network configuration and connections are applied automatically in {microshift-short} with the `microshift-networking` RPM during installation.
+* A cluster that uses the OVN-Kubernetes network plugin also runs Open vSwitch (OVS) on the node.
+* OVN-K configures OVS on the node to implement the declared network configuration.
+* Host physical interfaces are not bound by default to the OVN-K gateway bridge, `br-ex`. You can use standard tools for managing the default gateway.
```
Comment (kelbrown20) [inline microshift_networking/microshift-cni.adoc:14]: I'm not too sure what "standard tools" refers to here.

File: microshift_networking/microshift-cni.adoc [hunk 4]
```diff
@@ -6,7 +6,22 @@ include::_attributes/attributes-microshift.adoc[]
...
+* Changing the CNI is not supported on {microshift-short}.
+
+Using configuration files or custom scripts, you can configure the following networking settings:
+
+* You can use CIDR to route data packets to your devices. Routers retrieve the network and host address as specified by the CIDR suffix.
+* You can change the MTU value.
```
Comment (kelbrown20) [inline microshift_networking/microshift-cni.adoc:20]: I'm pretty sure thats right. 
```suggestion
* You can change the maximum transmission unit (MTU) value.
```
Unless you think this is inferred

File: microshift_networking/microshift-cni.adoc [hunk 5]
```diff
@@ -6,7 +6,22 @@ include::_attributes/attributes-microshift.adoc[]
...
+
+* Default network configuration and connections are applied automatically in {microshift-short} with the `microshift-networking` RPM during installation.
+* A cluster that uses the OVN-Kubernetes network plugin also runs Open vSwitch (OVS) on the node.
+* OVN-K configures OVS on the node to implement the declared network configuration.
+* Host physical interfaces are not bound by default to the OVN-K gateway bridge, `br-ex`. You can use standard tools on the host for managing the default gateway, such as the Network Manager CLI (`nmcli`).
+* Changing the CNI is not supported on {microshift-short}.
+
+Using configuration files or custom scripts, you can configure the following networking settings:
+
+* You can use CIDR to route data packets to your devices. Routers retrieve the network and host address as specified by the CIDR suffix.
```
Comment (pliurh) [inline microshift_networking/microshift-cni.adoc:19]: I don't quite get this sentence. Does it want to say that users can add custom routes to route the egress traffic?

File: microshift_networking/microshift-cni.adoc [hunk 6]
```diff
@@ -6,7 +6,22 @@ include::_attributes/attributes-microshift.adoc[]
...
+* You can change the maximum transmission unit (MTU) value.
+* You can configure firewall ingress and egress.
+* You can configure network policies, including ingress and egress rules.
```
Comment (pliurh) [inline microshift_networking/microshift-cni.adoc:22]: The network policies shall be defined in the Microshift cluster, instead of on the host.

File: modules/microshift-cni-customization-matrix.adoc [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+|===
+|Network feature|Availability|Customization status
+
+|Kubernetes network policy|Yes|Supported
+
+|Kubernetes network policy logs|Not available|N/A
+
+|Load balancing|Yes|Supported
+
+|Multicast|Yes|mDNS supported
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:19]: I don't remember we claim to support multicast in Microshift. mDNS and Multicast are two different features.

File: modules/microshift-cni-customization-matrix.adoc [hunk 8]
```diff
@@ -0,0 +1,46 @@
+
+:_mod-docs-content-type: REFERENCE
+[id="microshift-nw-ovn-kubernetes-matrix_{context}"]
+= {microshift-short} CNI feature matrix
+
+The following table summarizes the status of OVN-K networking features that are either present as defaults, supported for configuration, or excluded from the {microshift-short} service:
+
+.Default {microshift-short} OVN-K network plugin features and customization status
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:8]: Not all the network features in MicroShift are provided by OVN-K. https://docs.openshift.com/container-platform/4.14/networking/ovn_kubernetes_network_provider/about-ovn-kubernetes.html#nw-ovn-kubernetes-matrix_about-ovn-kubernetes

File: modules/microshift-cni-customization-matrix.adoc [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+[cols="50%,20%,30%",options="header"]
+|===
+|Network feature|Availability|Customization status
+
+|Kubernetes network policy|Yes|Supported
+
+|Kubernetes network policy logs|Not available|N/A
+
+|Load balancing|Yes|Supported
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:17]: This looks unclear to me. Does it mean the support to the LoadBalancer type of k8s Services?

File: modules/microshift-cni-customization-matrix.adoc [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+|Multicast|Yes|mDNS supported
+
+|Network proxies|Yes ^[1]^|CRI-O supported
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:21]: Shall it be listed as a network feature?

File: modules/microshift-cni-customization-matrix.adoc [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+|Hybrid networking|Not available|N/A
+
+|IPsec encryption for intra-cluster communication|Not available|N/A
+
+|IPv6|Not available|N/A
+|===
+
+1. There is no built-in transparent proxying of egress traffic in {microshift-short}. Egress must be manually configured.
+2. Setting up the firewalld service is supported by the OVN-K CNI.
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:43]: OVN-K does NOT support setting up firewalld.

File: microshift_networking/microshift-cni.adoc [hunk 12]
```diff
@@ -6,109 +6,64 @@ include::_attributes/attributes-microshift.adoc[]
...
-== Network performance optimizations
-By default, three performance optimizations are applied to OVS services to minimize resource consumption:
+Using configuration files or custom scripts, you can configure the following networking settings:
 
-* CPU affinity to `ovs-vswitchd.service` and `ovsdb-server.service`
-* `no-mlockall` to `openvswitch.service`
-* Limit handler and `revalidator` threads to `ovs-vswitchd.service`
+* You can use subnet CIDR ranges to allocate IP addresses to pods.
+* You can change the maximum transmission unit (MTU) value.
+* You can configure firewall ingress and egress.
```
Comment (pacevedom) [inline microshift_networking/microshift-cni.adoc:21]: Is this firewalld?

File: microshift_networking/microshift-cni.adoc [hunk 13]
```diff
@@ -6,109 +6,64 @@ include::_attributes/attributes-microshift.adoc[]
...
+* You can define network policies in the {microshift-short} cluster, including ingress and egress rules.
```
Comment (pacevedom) [inline microshift_networking/microshift-cni.adoc:22]: We might need a note around ingress network policies, as not all service types allow for it, and for the one that does so it needs another setting so that it works. Where would that fit?

File: modules/microshift-cni-customization-matrix.adoc [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+|===
+|Network feature|Availability|Customization supported
+
+|Advertise address|Yes|Yes ^[1]^
+
+|Kubernetes network policy|Yes|Yes
+
+|Kubernetes network policy logs|Not available|N/A
+
+|Load balancing|Yes|Yes
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:19]: This looks unclear to me. Does it mean the support to the LoadBalancer type of k8s Services?

File: modules/microshift-nw-topology.adoc [hunk 15]
```diff
@@ -0,0 +1,46 @@
+// Module included in the following assemblies:
+//
+// * microshift_networking/microshift-cni.adoc
+
+:_mod-docs-content-type: CONCEPT
+[id="microshift-network-topology_{context}"]
+= Network topology
+OVN-Kubernetes provides an overlay-based networking implementation. This overlay includes an OVS-based implementation of Service and NetworkPolicy. The overlay network uses the Geneve (Generic Network Virtualization Encapsulation) tunnel protocol. The pod maximum transmission unit (MTU) for the Geneve tunnel is set to a smaller value than the MTU of the physical interface on the host. This smaller MTU makes room for the required information that is added to the tunnel header before it is transmitted.
```
Comment (pliurh) [inline modules/microshift-nw-topology.adoc:8]: This statement is inaccurate. Please refer to https://github.com/openshift/microshift/blob/f0d95cf39d316f396cea7af4c329d8de9389f579/docs/contributor/network/default_cni_plugin.md?plain=1#L68

File: modules/microshift-cni-customization-matrix.adoc [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+
+|Multicast DNS|Yes|Yes ^[2]^
+
+|Network proxies|Yes ^[3]^|CRI-O
```
Comment (pliurh) [inline modules/microshift-cni-customization-matrix.adoc:23]: I don't know what it means. @pacevedom Do you have any idea?

File: modules/microshift-config-yaml.adoc [hunk 17]
```diff
@@ -8,63 +8,11 @@
...
-<5> The name of the node. The default value is the hostname.
-<6> The IP address of the node. The default value is the IP address of the default route.
-<7> A string that specifies the IP address from which the API server is advertised to members of the cluster. The default value is calculated based on the address of the service network.
-<8> Subject Alternative Names for API server certificates.
-<9> Log verbosity. Valid values for this field are `Normal`, `Debug`, `Trace`, or `TraceAll`.
-
 [id="microshift-yaml-custom_{context}"]
 == Custom settings
-To create custom configurations, you must create a `config.yaml` file in the `/etc/microshift/` directory, then change any settings that should override the defaults before starting or restarting {microshift-short}.
+To create custom configurations, you must create a `config.yaml` file in the `/etc/microshift/` directory, then change any settings that are expected to override the defaults before starting or restarting {microshift-short}.
```
Comment (lpettyjo) [inline modules/microshift-config-yaml.adoc:13]: Change to "...directory, and then change any..."

File: modules/microshift-default-settings.adoc [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+// * microshift_configuring/microshift-using-config-tools.adoc
+// * microshift_networking/
+
+:_mod-docs-content-type: CONCEPT
+[id="microshift-yaml-default_{context}"]
+= Default settings
+
+If you do not create a `config.yaml` file, default values are used. The following example shows the default configuration settings.
+
+* You can use the following command to see the default values:
```
Comment (lpettyjo) [inline modules/microshift-default-settings.adoc:12]: Can you shorten this to "Use the following command..."? Are there other commands to view the default values? If not, I'd use my suggestion.

File: modules/microshift-nw-components-svcs.adoc [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+OVN containers::
+Two OVN-Kubernetes daemon sets are rendered and applied by {microshift-short}.
+
+* *ovnkube-master*
+Includes the `northd`, `nbdb`, `sbdb` and `ovnkube-master` containers.
+
+* *ovnkube-node*
+The ovnkube-node includes the OVN-Controller container.
++
+After {microshift-short} boots, the OVN-Kubernetes daemon sets are deployed in the `openshift-ovn-kubernetes` namespace.
```
Comment (lpettyjo) [inline modules/microshift-nw-components-svcs.adoc:25]: Can you substitute "starts" for "boots"? If so, it is preferred if you are not losing technical accuracy.

File: modules/microshift-nw-topology.adoc [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+
+OVN gateway router::
+A virtual router named `GR_<node-name>`, also known as the external gateway router.
+
+OVN external switch::
+A virtual switch named `ext_<node-name>.`
+
+[id="microshift-description-connections-network-topology_{context}"]
+== Description of the connections in the network topology figure
+* The north-south traffic between the network service and the OVN external switch `ext_microshift-dev`, is provided through the host kernel by the gateway bridge `br-ex`.
```
Comment (lpettyjo) [inline modules/microshift-nw-topology.adoc:38]: Remove comma.

**Review by lpettyjo — COMMENTED** [review]
Otherwise, LGTM!

File: modules/microshift-nw-advertise-address.adoc [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+= Configuring the advertise address network flag
+
+The `apiserver.advertiseAddress` flag specifies the IP address on which to advertise the API server to members of the cluster. This address must be reachable by the cluster. You can set a custom IP address here, but you must also add the IP address to a host interface. Customizing this parameter preempts {microshift-short} from adding a default IP address to the `br-ex` network interface.
+
+[IMPORTANT]
+====
+If you customize the `advertiseAddress` IP address, make sure it is reachable by the cluster when {microshift-short} starts by adding the IP address to a host interface.
+====
+
+If unset, the default value is `10.44.0.0/32`. It will be set to the next immediate subnet after the service network. For example, when the service network is `10.43.0.0/16`, the `advertiseAddress` is set to `10.44.0.0/32`.
```
Comment (pacevedom) [inline modules/microshift-nw-advertise-address.adoc:16]: I would remove the first part about the default value being '10.44.0.0/32', as it is explained right after that. There is also another example too.

File: modules/microshift-cni-customization-matrix.adoc [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+|Hardware offloading|Not available|N/A
+
+|Hybrid networking|Not available|N/A
+
+|IPsec encryption for intra-cluster communication|Not available|N/A
+
+|IPv6|Not available|N/A
+|===
+
+1. If unset, the default value is `10.44.0.0/32`. It will be set to the next immediate subnet after the service network.
```
Comment (pacevedom) [inline modules/microshift-cni-customization-matrix.adoc:44]: I would write the example after the description. If users change the service network but dont change the advertise address, default value is not 10.44.0.0/32.

File: modules/microshift-cni-customization-matrix.adoc [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+
+|Firewall|No ^[4]^|Yes
+
+|Hardware offloading|Not available|N/A
+
+|Hybrid networking|Not available|N/A
+
+|IPsec encryption for intra-cluster communication|Not available|N/A
+
+|IPv6|Not available|N/A
```
Comment (anuragthehatter) [inline modules/microshift-cni-customization-matrix.adoc:41]: Since we are explicitly defining whats supported or not. Does Mentioning IPv6 not available implies that Dual stack is not supported? 

rest technicalities looks good as most of the part has been reviewed in last releases

Comment (anuragthehatter) [reply modules/microshift-cni-customization-matrix.adoc:41]: Or it probably implies that IPv6 is not available in terms of dual stack and single stack IPv6

Comment (ShaunaDiaz) [reply modules/microshift-cni-customization-matrix.adoc]: Clarified in a footnote that IPv6 is not supported in any config.
